src/java.base/share/classes/sun/net/www/protocol/http/AuthCacheImpl.java

```diff
@@ -37,15 +40,55 @@ public class AuthCacheImpl implements AuthCache {
     // in this class, so there is no need to convert this class to using
     // java.util.concurrent.locks
     HashMap<String,LinkedList<AuthCacheValue>> hashtable;
+    final Cleaner cleaner;
 
     public AuthCacheImpl () {
         hashtable = new HashMap<String,LinkedList<AuthCacheValue>>();
+        cleaner = Cleaner.create();
     }
 
     public void setMap (HashMap<String,LinkedList<AuthCacheValue>> map) {
         hashtable = map;
     }
 
+    // call to register Authenticator with Cleaner.
+
+    public void registerAuthenticator(Authenticator auth) {
+        cleaner.register(auth, new CleanerAction(AuthenticatorKeys.getKey(auth)));
+    }
+
+    // used for testing
+    public int mapSize() {
+        AtomicInteger count = new AtomicInteger();
+        hashtable.forEach((k,v) -> {
+            count.addAndGet(v == null ? 0: v.size());
+        });
+        return count.get();
+    }
+
+    // Cleaner action run to remove all entries whose key ends with
+    // ;auth=authkey
+    // ie. all entries belonging to the given Authenticator
+
+    class CleanerAction implements Runnable {
+        private final String authkey;
+
+        CleanerAction(String authkey) {
+            this.authkey = ";auth=" + authkey;
+        }
+
+        public void run() {
+            synchronized(AuthCacheImpl.this) {
+                hashtable.forEach((String key,
+                                   LinkedList<AuthCacheValue> list) -> {
+                    if (key.endsWith(authkey)) {
+                        hashtable.remove(key);
```

Comment:
won't this throw a `ConcurrentModificationException`?

Reply:
```
jshell> var map = Map.of(1,1,2,2,3,3,4,4)
map ==> {1=1, 2=2, 3=3, 4=4}

jshell> var map2 = new HashMap<Integer, Integer>();
map2 ==> {}

jshell> map2.putAll(map)

jshell> map2.forEach((k, v) -> map2.remove(k));
|  Exception java.util.ConcurrentModificationException
|        at HashMap.forEach (HashMap.java:1432)
|        at (#4:1)

jshell> var map3 = new ConcurrentHashMap<Integer, Integer>();
map3 ==> {}

jshell> map3.putAll(map)

jshell> map3.forEach((k, v) -> map3.remove(k));

jshell> map3.isEmpty()
$10 ==> true
```

Reply:
Right. It could throw CME, but odd that it doesn't in the test. I'd prefer to use an Iterator than change to a ConcurrentHashMap.